File: js/src/SafeFetch.js
/* 
    Written by Prizzle#4655,
    Method discovered by Milkyway and Sam,
    all rights reserved.
*/

import fetch from "node-fetch";

export default class SafeFetch {
	#maxAttemps = 50;
	#log = (text) => console.log(`[${new Date().toISOString()}] ${text}`);
	#counter = 0;

	async #generateSafeFetchToken(url) {
		try {
			this.#log(`Generating safe fetch token for ${url}`);
			let fetchRaw = await fetch(
				`https://docs.google.com/gview?url=${url}`,
				{
					headers: {
						accept: "*/*",
						"accept-language":
							"de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7",
						"sec-ch-ua":
							'" Not A;Brand";v="99", "Chromium";v="101", "Google Chrome";v="101"',
						"sec-ch-ua-mobile": "?0",
						"sec-ch-ua-platform": '"Windows"',
						"sec-fetch-dest": "empty",
						"sec-fetch-mode": "cors",
						"sec-fetch-site": "same-origin",
						Referer: `https://docs.google.com/gview?url=${url}`,
						"Referrer-Policy": "strict-origin-when-cross-origin",
					},
				}
			);
			if (!fetchRaw.ok || fetchRaw.status == 204) {
				this.#log("Could not generate safe fetch token...");
				return {
					success: false,
					message: `Request was not OK or just empty (${fetchRaw.status})`,
					content: fetchRaw,
				};
			}

			if (fetchRaw.redirected) {
				return {
					message: `Request was unexpectedly redirected to ${fetchRaw.url}`,
					success: false,
					content: fetchRaw,
				};
			}

			let fetchText = await fetchRaw.text();

			//console.log(fetchText);

			if (!fetchText) {
				throw new Error("Fetch got no text");
			}

			let id = fetchText
				?.split(`text?id\\u003d`)?.[1]
				?.split("\\u0026authuser")?.[0]
				?.split('"')[0];

			if (!id) {
				throw new Error("Could not find id");
			}

			return {
				success: true,
				message: "Successfully fetched id",
				content: id,
			};
		} catch (err) {
			return {
				success: false,
				message: `${err.name}, ${err.message}`,
				content: "",
			};
		}
	}
	async #renderSafeFetch(safeFetchId) {
		try {
			let fetchRaw = await fetch(
				`https://docs.google.com/viewerng/text?id=${safeFetchId}&page=0`,
				{
					/* agent: this.#proxyAgent, */
					headers: {
						accept: "*/*",
						"accept-language":
							"de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7",
						"sec-ch-ua":
							'" Not A;Brand";v="99", "Chromium";v="101", "Google Chrome";v="101"',
						"sec-ch-ua-mobile": "?0",
						"sec-ch-ua-platform": '"Windows"',
						"sec-fetch-dest": "empty",
						"sec-fetch-mode": "cors",
						"sec-fetch-site": "same-origin",
					},
				}
			);

			if (!fetchRaw.ok) {
				this.#log("Could not generate safe fetch view...");
				return {
					success: false,
					message: `Request was not OK (${fetchRaw.status})`,
					content: fetchRaw,
				};
			}

			if (fetchRaw.redirected) {
				return {
					message: `Request was unexpectedly redirected to ${fetchRaw.url}`,
					success: false,
					content: fetchRaw,
				};
			}

			let fetchText = await fetchRaw.text();
			let mutateText = fetchText.split(`\n`)[1];
			let fetchJSON = JSON.parse(mutateText);

			if (
				fetchJSON.mimetype == "text/html" &&
				typeof DOMParser != "undefined"
			) {
				let resultHTML = new DOMParser().parseFromString(
					fetchJSON.data,
					"text/html"
				);
				return {
					message:
						"Successfully requested URL and processed content to HTML",
					success: true,
					content: resultHTML,
				};
			} else if (fetchJSON.mimetype == "application/json") {
				try {
					let resultJSON = JSON.parse(fetchJSON.data);
					return {
						message:
							"Successfully requested URL and processed content to JSON",
						success: true,
						content: resultJSON,
					};
				} catch (err) {
					let text = fetchJSON.data;

					return {
						message:
							"Successfully requested URL but failed to parse JSON",
						success: true,
						content: text,
					};
				}
			} else {
				return {
					message: "Successfully requested URL",
					success: true,
					content: fetchJSON,
				};
			}
		} catch (err) {
			return {
				success: false,
				message: `${err.name}, ${err.message}`,
				content: null,
			};
		}
	}
	async get(url) {
		// Attempt to generate a token
		// for the next request
		let tokenGenerationRequest;
		let attempts = 0;
		while (true) {
			tokenGenerationRequest = await this.#generateSafeFetchToken(url);
			attempts++;
			if (tokenGenerationRequest.success) break;
			if (attempts > this.#maxAttemps)
				return {
					message:
						"Exceeded maximum safe fetch token generation attempts",
					success: false,
					content: null,
				};
		}

		// Render your requested content
		// with the generated token
		let renderedContent = await this.#renderSafeFetch(
			tokenGenerationRequest["content"]
		);
		this.#log(
			`[${this.#counter}] Successfully rendered safe fetch content`
		);
		this.#counter++;
		return renderedContent;
	}

	setAttemptLimit(number = this.#maxAttemps) {
		if (!number) return false;
		if (isNaN(parseInt(number))) return false;
		this.#maxAttemps = number;
	}
}

File: go/src/safeFetch/safeFetch.go
package safeFetch

import (
	"fmt"
	"io/ioutil"
	"strings"
	"time"

	http "github.com/bogdanfinn/fhttp"
	tls "github.com/bogdanfinn/tls-client"
)

type SafeFetch struct {
	client      tls.HttpClient
	maxAttempts int
	counter     int
}

func NewSafeFetch(max int) *SafeFetch {
	options := []tls.HttpClientOption{
		tls.WithTimeout(10),
		tls.WithClientProfile(tls.Chrome_110),
	}

	client, err := tls.NewHttpClient(nil, options...)
	if err != nil {
		panic(err)
	}

	return &SafeFetch{
		maxAttempts: max,
		client:      client,
		counter:     0,
	}
}

func (s *SafeFetch) SetMaxAttempts(maxAttempts int) {
	s.maxAttempts = maxAttempts
}

func (s *SafeFetch) generateToken(url string) (string, error) {
	req, err := http.NewRequest("GET", fmt.Sprintf("https://docs.google.com/gview?url=%s", url), nil)
	if err != nil {
		return "", err
	}

	req.Header = http.Header{
		"accept":             []string{"*/*"},
		"accept-language":    []string{"de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7"},
		"sec-ch-ua":          []string{"\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"101\", \"Google Chrome\";v=\"101\""},
		"sec-ch-ua-mobile":   []string{"?0"},
		"sec-ch-ua-platform": []string{"\"Windows\""},
		"sec-fetch-dest":     []string{"empty"},
		"sec-fetch-mode":     []string{"cors"},
		"sec-fetch-site":     []string{"same-origin"},
		"Referer":            []string{fmt.Sprintf("https://docs.google.com/gview?url=%s", url)},
		"Referer-Policy":     []string{"strict-origin-when-cross-origin"},

		http.HeaderOrderKey: []string{
			"accept",
			"accept-language",
			"sec-ch-ua",
			"sec-ch-ua-mobile",
			"sec-ch-ua-platform",
			"sec-fetch-dest",
			"sec-fetch-mode",
			"sec-fetch-site",
			"Referer",
			"Referer-Policy",
		},
	}

	resp, err := s.client.Do(req)
	if err != nil {
		return "", err
	}

	if resp.StatusCode != 200 {
		return "", fmt.Errorf("Status code is not 200")
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	split1 := strings.Split(string(body), `text?id\u003d`)

	if len(split1) < 2 {
		return "", fmt.Errorf("Could not find token step 1")
	}

	split2 := strings.Split(split1[1], `\\u0026authuser`)

	if len(split2) < 1 {
		return "", fmt.Errorf("Could not find token step 2")
	}

	split3 := strings.Split(split2[0], `"`)

	if len(split3) < 1 {
		return "", fmt.Errorf("Could not find token step 3")
	}

	id := split3[0]

	return id, nil
}

func (s *SafeFetch) render(token string) (string, error) {
	req, err := http.NewRequest("GET", fmt.Sprintf("https://docs.google.com/viewerng/text?id=%s&page=0", token), nil)
	if err != nil {
		return "", err
	}

	req.Header = http.Header{
		"accept":             []string{"*/*"},
		"accept-language":    []string{"de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7"},
		"sec-ch-ua":          []string{"\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"101\", \"Google Chrome\";v=\"101\""},
		"sec-ch-ua-mobile":   []string{"?0"},
		"sec-ch-ua-platform": []string{"\"Windows\""},
		"sec-fetch-dest":     []string{"empty"},
		"sec-fetch-mode":     []string{"cors"},
		"sec-fetch-site":     []string{"same-origin"},

		http.HeaderOrderKey: []string{
			"accept",
			"accept-language",
			"sec-ch-ua",
			"sec-ch-ua-mobile",
			"sec-ch-ua-platform",
			"sec-fetch-dest",
			"sec-fetch-mode",
			"sec-fetch-site",
		},
	}

	resp, err := s.client.Do(req)
	if err != nil {
		return "", err
	}

	if resp.StatusCode != 200 {
		return "", fmt.Errorf("Status code is not 200")
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	content := strings.Split(string(body), "\n")

	if len(content) < 2 {
		return "", fmt.Errorf("Could not find content")
	}

	return content[1], nil
}
func (s *SafeFetch) Get(url string) (string, error) {
	fmt.Println(fmt.Sprintf("[%s] Generating token for %s", time.Now().Format("2006-01-02 15:04:05"), url))

	var token string

	for i := 0; i < s.maxAttempts; i++ {
		var err error

		token, err = s.generateToken(url)
		if err != nil {
			if err.Error() == "Status code is not 200" {
				fmt.Println(fmt.Sprintf("[%s] Failed to generate token, retrying...", time.Now().Format("2006-01-02 15:04:05")))
				continue
			}

			return "", err
		}

		if token != "" {
			break
		}
	}

	if token == "" {
		return "", fmt.Errorf(fmt.Sprintf("[%s] Maximum token generation attempts reached", time.Now().Format("2006-01-02 15:04:05")))
	}

	render, err := s.render(token)
	if err != nil {
		return "", err
	}

	s.counter++
	fmt.Println(fmt.Sprintf("[%s][%d] Successfully rendered document", time.Now().Format("2006-01-02 15:04:05"), s.counter))

	return render, nil
}
